### ISSUE TITLE
Support for remapping Kotlin (and others) source code

### ISSUE THREAD

**Issue by LoganDark** [FabricMC/fabric-loom#286]
Some related topics have been discussed in issues #239 and #129 (and by extension #126), but this has not been addressed in a dedicated issue.

Loom does not seem to remap Kotlin sources. They are not remapped to intermediary in the RemapSourcesJarTask, and when loading the sources jar (for example from a Gradle dependency) intermediaries are not remapped to your current mappings.

So for example if a project uses mojmaps for whatever reason, by viewing its non-Java sources (through your IDE) you would be viewing Mojang names, not Yarn names. Your IDE would get upset because it wouldn't be able to find any of the classes, and you would have to build in your head an association between Mojang names and Yarn names, which is not a good thing if you are trying to stay pure. I personally would like to contribute to Yarn in the future, I would like to keep my head clean.

It has been brought up before that "It's not Fabric's job to support non-Java languages". I think this should be reconsidered for a few reasons:
- Clearly, they are popular enough to warrant their own dedicated library mods hosted under the FabricMC repository. (for Kotlin and Groovy at least)
- Many members of the FabricMC organization itself even use non-Java languages (mainly Kotlin). Examples: [2xsaiko](https://github.com/2xsaiko/HCTM), [Juuxel](https://github.com/Juuxel/Adorn)
- Development in these languages tends to be nicer than development in Java. Fabric already recognizes this through official support of their standard libraries. Why not support them at the source code level too?

Support for Kotlin (and others) would make my development experience nicer because I could then:
- Safely view sources knowing that they have been remapped properly
- Safely write sources knowing that they will be remapped properly, and that other developers will see proper names suited to their mappings

**Comment by modmuss50** [comment]
I’d love to be able to do this, however I’m not aware of any kotlin source remapping tools. 

We use mercury to remap java code with great success.

**Comment by LoganDark** [comment]
Yeah... I think making your own Kotlin source remapping tool is definitely out of the scope of Fabric. But if one already exists, adding support wouldn't be.

**Comment by modmuss50** [comment]
I agree, I’ll leave this open in case anyone can find or make one

**Comment by LoganDark** [comment]
https://github.com/ReplayMod/remap looks relevant, but it needs more research to see if it would even be viable to have support added to Loom. Also I don't know if it's a source remapper _for_ Kotlin or a source remapper _in_ Kotlin. It says it uses an embedded IntelliJ IDEA, so it might support all refactoring options that IDEA does (hello free Groovy+Scala support), but that may be a lot of bloat for a source remapper.

Issues also suggest it supports Yarn already. But Replay Mod is Java, not Kotlin.

**Comment by LoganDark** [comment]
@modmuss50 Would you mind looking into ReplayMod/remap and seeing if it's usable? I don't know much about remapping so I can't do it myself :/

**Comment by i509VCB** [comment]
Unsure we would want to support ReplayMod's thing, but we could introduce an api to remap sources possibly?

**Comment by LoganDark** [comment]
@i509VCB also totally welcome, because that means someone's probably gonna invent the kotlin remapping for you :P

**Comment by ghostflyby** [comment]
How about using https://github.com/openrewrite/rewrite for remapping all sources?

It supports both Java and Kotlin.

**Comment by ExpensiveKoala** [comment]
I did a little bit of digging into this out of curiosity and I want to share my findings.

Loom relies on https://github.com/CadixDev/Mercury to handle mapping migration. But mercury relies on Java AST (Abstract Syntax Tree) parsing. In order to support Kotlin, mercury would have to be forked to use Kotlin AST or similar. Though I don't know if a Kotlin AST library exists that can also handle rewriting. (The Java AST Mercury uses is developed by eclipse for their IDE so it includes rewriting capabilities).
The way it works is find source files -> parse into AST -> apply mapping conversions -> output source files based on the AST

The most up to date Kotlin AST library https://github.com/kotlinx/ast doesn't have any features built-in for outputting source based on AST. It is purely for parsing.

Potentially one can use KotlinPoet https://github.com/square/kotlinpoet to output the sources based on the parsed AST, but I'm not entirely sure how feature complete that is...

Alternatively, https://github.com/cretz/kastree exists which is another Kotlin AST library that includes writing sources based on the AST, but it hasn't seen a commit in 6 years so it probably doesn't support some of the more recent Kotlin language features.

An option just for migrateMappings would be to create an intellij plugin which can use the PSI for code parsing and refactoring to achieve the migrateMappings functionality.
